mono/metadata/marshal-ilgen.c

```diff
@@ -5483,7 +5483,7 @@ emit_marshal_object_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 				}
 
 				mono_mb_emit_byte (mb, CEE_STIND_REF);
-			} else {
+			} else if (t->attrs & PARAM_ATTRIBUTE_OUT) {
```

Comment:
It is both unnecessary and functionally wrong to marshal back the buffer contents. TBD: Check other places if the check is missing somewhere else too.